src/graphql.tsx

```diff
@@ -276,14 +300,27 @@ export default function graphql(
         return opts;
       };
 
-      calculateResultProps(result) {
-        let name = this.type === DocumentType.Mutation ? 'mutate' : 'data';
-        if (operationOptions.name) name = operationOptions.name;
-
-        const newResult = { [name]: result, ownProps: this.props };
-        if (mapResultToProps) return mapResultToProps(newResult);
+      calculateResultProps<T>(result: T) {
+        // Ugly, but the hope is to allow typescript to do control-flow analysis
+        // to determine if `data` or `mutate` are the keys
+        if (operationOptions.name != null) {
+          let name = operationOptions.name;
+          const newResult = { [name]: result, ownProps: this.props };
+          // Prevents us inferring useful type information :/
+          if (mapResultToProps) return mapResultToProps<typeof newResult, { [i: string]: any }>(newResult);
+
+          return { [name]: defaultMapResultToProps(result) };
+        } else if (this.type === DocumentType.Mutation) {
```

Comment:
This is another cash that prevents us from deducing the type information easily. Since the DocumentType is determined at runtime, we have to dynamically assign the key as either `mutate` or `data`. There is potentially a way to solve this via [Type Guards](https://basarat.gitbooks.io/typescript/content/docs/types/typeGuard.html), but I am not 100% sure it will work as expected.

Another (possibly better in the long term) solution would be to separate the `graphql` function into two functions: `query` and `mutate`. This way, a lot less work is spent determining things at runtime; and we can make better guarantees about the types.